modules/shell/command.go

```diff
@@ -96,84 +106,85 @@ func runCommandAndStoreOutputE(t testing.TestingT, command Command, storedStdout
 
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	stderr, err := cmd.StderrPipe()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	err = cmd.Start()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	if err := readStdoutAndStderr(t, command.Logger, stdout, stderr, storedStdout, storedStderr); err != nil {
-		return err
-	}
-
-	if err := cmd.Wait(); err != nil {
-		return err
+	output, err := readStdoutAndStderr(t, command.Logger, stdout, stderr)
+	if err != nil {
+		return output, err
 	}
 
-	return nil
+	return output, cmd.Wait()
 }
 
 // This function captures stdout and stderr into the given variables while still printing it to the stdout and stderr
 // of this Go program
-func readStdoutAndStderr(t testing.TestingT, log *logger.Logger, stdout, stderr io.ReadCloser, storedStdout, storedStderr *[]string) error {
+func readStdoutAndStderr(t testing.TestingT, log *logger.Logger, stdout, stderr io.ReadCloser) (*output, error) {
+	out := newOutput()
 	stdoutReader := bufio.NewReader(stdout)
 	stderrReader := bufio.NewReader(stderr)
 
 	wg := &sync.WaitGroup{}
-	mutex := &sync.Mutex{}
 
 	wg.Add(2)
 	var stdoutErr, stderrErr error
 	go func() {
 		defer wg.Done()
-		stdoutErr = readData(t, log, stdoutReader, mutex, storedStdout)
+		stdoutErr = readData(t, log, stdoutReader, out.stdout)
 	}()
 	go func() {
 		defer wg.Done()
-		stderrErr = readData(t, log, stderrReader, mutex, storedStderr)
+		stderrErr = readData(t, log, stderrReader, out.stderr)
 	}()
 	wg.Wait()
 
 	if stdoutErr != nil {
-		return stdoutErr
+		return out, stdoutErr
 	}
 	if stderrErr != nil {
-		return stderrErr
+		return out, stderrErr
 	}
 
-	return nil
+	return out, nil
 }
 
-func readData(t testing.TestingT, log *logger.Logger, reader *bufio.Reader, mutex *sync.Mutex, allOutput *[]string) error {
+func readData(t testing.TestingT, log *logger.Logger, reader *bufio.Reader, writer io.StringWriter) error {
 	var line string
 	var err error
 	for {
 		line, err = reader.ReadString('\n')
 
-		// remove newline, our output is in a slice,
-		// one element per line.
-		line = strings.TrimSuffix(line, "\n")
-
-		// only return early if the line does not have
-		// any contents. We could have a line that does
-		// not not have a newline before io.EOF, we still
-		// need to add it to the output.
-		if len(line) == 0 && err == io.EOF {
-			break
+		// separate block so that we do not overwrite the err variable
+		// that we need afterwards to check if we're done
+		{
+			// remove newline, our output is in a slice,
+			// one element per line.
+			line = strings.TrimSuffix(line, "\n")
+
+			// only return early if the line does not have
+			// any contents. We could have a line that does
+			// not not have a newline before io.EOF, we still
+			// need to add it to the output.
+			if len(line) == 0 && err == io.EOF {
+				break
+			}
+
+			log.Logf(t, line)
+			if _, err := writer.WriteString(line); err != nil {
```

Comment:
Since there's no more mutex, we're assuming this writer is goroutine-safe?

Reply:
No; as they don't need to be (unless they write to the same underlying store).
As long as the caller does not call this method in parallel with the same writer, we do not need a goroutine-safe one.

In this case, `output.merged` is protected by a mutex. However, only because of the structure of it, there's no relation to this function here.

Reply:
> As long as the caller does not call this method in parallel with the same writer, we do not need a goroutine-safe one.

And I guess since `readData` is a private method, you've gone through all the existing callers to make sure they aren't using the same writer anywhere? 

Either way, it's worth updating the comment on `readData` that the method is NOT go-routine safe, so you should not re-use writers.